infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java

```diff
@@ -179,24 +237,46 @@ public Builder setAttributes(Map<String, String> attributes) {
       return this;
     }
 
-    public Builder setType(String type) {
+    protected Builder setType(String type) {
       this.type = type;
       return this;
     }
 
     public KubernetesEnvironment build() {
-      KubernetesEnvironment kubernetesEnvironment =
-          new KubernetesEnvironment(
-              internalRecipe,
-              machines,
-              warnings,
-              pods,
-              services,
-              ingresses,
-              pvcs,
-              secrets,
-              configMaps);
-      kubernetesEnvironment.setAttributes(attributes);
+      KubernetesEnvironment kubernetesEnvironment;
+      if (internalEnvironment == null) {
+        kubernetesEnvironment =
+            new KubernetesEnvironment(
+                internalRecipe,
+                machines,
+                warnings,
+                pods,
+                services,
+                ingresses,
+                pvcs,
+                secrets,
+                configMaps);
+      } else {
+        kubernetesEnvironment =
+            new KubernetesEnvironment(
+                internalEnvironment, pods, services, ingresses, pvcs, secrets, configMaps);
+        if (internalRecipe != null) {
+          kubernetesEnvironment.setRecipe(internalRecipe);
```

Comment:
I think that setting these values in InternalEnvironment object would simplify this code and builder setters

Reply:
Do you mean that we should create empty `Environment` and instead of holding own fiends in `Builder` just set fill values in `Environment` instance? If yes, then I agree that it would simplify this code BUT I see one issue here: build won't be failed when a new field is added to `Environment` while the current approach fail build because it uses Constructor with all values specified.

Reply:
I mean that we could use setters of parent InternalEnvironment in the build setters and not having additional fields here and not have them in the constructor and not check nullability here

Reply:
Reworked. Code looks simpler with proposed changes. Thanks